### PR TITLE
fix(rome_js_formatter): line break before assignment

### LINKED ISSUE #2417: Function and class expression assignment
Rome breaks any `function` or `class` expression assignment over at least two lines if their bodies aren't empty. Prettier formats them on a single line (as one would expect). [Playground](https://play.rome.tools/?lineWidth=120&indentStyle=space&quoteStyle=double&indentWidth=2&typescript=true&jsx=false&sourceType=module#LwAvACAAQQBzAHMAaQBnAG4AaQBuAGcAIAB0AG8AIABmAHUAbgBjAHQAaQBvAG4ALwBjAGwAYQBzAHMAZQBzACAAdwBpAHQAaAAgAGEAbgAgAGUAbQBwAHQAeQAgAHkAaQBlAGwAZABzACAAdABoAGUAIABkAGUAcwBpAHIAZQBkACAAZgBvAHIAbQBhAHQAdABpAG4AZwAKAGEAIAA9ACAAZgB1AG4AYwB0AGkAbwBuACgAKQAgAHsAfQAKAGEAIAA9ACAAYwBsAGEAcwBzACAAewB9AAoACgAvAC8AIABBAHMAcwBpAGcAbgBpAG4AZwAgAHQAbwAgAGEAIABmAHUAbgBjAHQAaQBvAG4ALwBjAGwAYQBzAHMAIAB3AGkAdABoACAAYQAgAG4AbwBuAC0AZQBtAHAAdAB5ACAAYgBvAGQAeQAgAGcAcgBvAHUAcABzACAAdABoAGUAIAB3AGgAbwBsAGUAIABhAHMAcwBpAGcAbgBtAGUAbgB0ACAAZQB4AHAAcgBlAHMAcwBpAG8AbgAuAAoAYQAgAD0AIABmAHUAbgBjAHQAaQBvAG4AIAAoACkAIAB7AAoAIAAgAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoADEAKQAKAH0AOwAKAAoAZQB4AHQAcgBhAEwAbwBuAGcATABlAGYAdABIAGEAbgBkAFMAaQBkAGUATABlAHQAcwBTAGUAZQBIAG8AdwBQAHIAZQB0AHQAaQBlAHIAQgByAGUAYQBrAHMAVABoAGEAdABBAGMAcgBvAHMAcwBNAHUAbAB0AGkAcABsAGUATABpAG4AZQBzAFMAdABpAGwAbABOAG8AdABMAG8AbgBnAEUAbgBvAHUAZwBoAE0AdQBzAHQAQgBlAEwAbwBuAGcAZQByAEEAbgBkAEwAbwByAGEAIAA9ACAAZgB1AG4AYwB0AGkAbwBuACAAKAApACAAewAKACAAIABjAG8AbgBzAG8AbABlAC4AbABvAGcAKAAxACkACgB9ADsACgAKAGEAIAA9ACAAZgB1AG4AYwB0AGkAbwBuACAAZQB4AHQAcgBhAEwAbwBuAGcATABlAGYAdABIAGEAbgBkAFMAaQBkAGUATABlAHQAcwBTAGUAZQBIAG8AdwBQAHIAZQB0AHQAaQBlAHIAQgByAGUAYQBrAHMAVABoAGEAdABBAGMAcgBvAHMAcwBNAHUAbAB0AGkAcABsAGUATABpAG4AZQBzAFMAdABpAGwAbABOAG8AdABMAG8AbgBnAEUAbgBvAHUAZwBoAE0AdQBzAHQAQgBlAEwAbwBuAGcAZQByAEEAbgBkAGEAcgBhACAAKAApACAAewAKACAAIABjAG8AbgBzAG8AbABlAC4AbABvAGcAKAAxACkACgB9ADsACgAKAGEAIAA9ACAAYwBsAGEAcwBzACAAewAKACAAIABwAHIAbwBwADsACgB9AAoACgAvAC8AIABGAHUAbgBjAHQAaQBvAG4AIAB3AGkAdABoACAAbgBvAG4ALQBlAG0AcAB0AHkAIABiAG8AZAB5ACAAaQBzACAAcAByAGkAbgB0AGUAZAAgAG8AbgAgAHQAaABlACAAbgBlAHgAdAAgAGwAaQBuAGUACgBBAC4AcAByAG8AdABvAHQAeQBwAGUALgBmACAAPQAgAGYAdQBuAGMAdABpAG8AbgAoACkAIAB7AAoAIAAgAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoADEAKQAKAH0ACgAKAA==)

## Input

```ts
// Assigning to function/classes with an empty yields the desired formatting
a = function() {}
a = class {}

// Assigning to a function/class with a non-empty body groups the whole assignment expression.
a = function () {
  console.log(1)
};

extraLongLeftHandSideLetsSeeHowPrettierBreaksThatAcrossMultipleLinesStillNotLongEnoughMustBeLongerAndLora = function () {
  console.log(1)
};

a = function extraLongLeftHandSideLetsSeeHowPrettierBreaksThatAcrossMultipleLinesStillNotLongEnoughMustBeLongerAndara () {
  console.log(1)
};

a = class {
  prop;
}

// Function with non-empty body is printed on the next line
A.prototype.f = function() {
  console.log(1)
}
```

## Prettier

```ts
// Assigning to function/classes with an empty yields the desired formatting
a = function () {};
a = class {};

// Assigning to a function/class with a non-empty body groups the whole assignment expression.
a = function () {
  console.log(1);
};

extraLongLeftHandSideLetsSeeHowPrettierBreaksThatAcrossMultipleLinesStillNotLongEnoughMustBeLongerAndLora =
  function () {
    console.log(1);
  };

a =
  function extraLongLeftHandSideLetsSeeHowPrettierBreaksThatAcrossMultipleLinesStillNotLongEnoughMustBeLongerAndara() {
    console.log(1);
  };

a = class {
  prop;
};

// Function with non-empty body is printed on the next line
A.prototype.f = function () {
  console.log(1);
};

```

## Rome

```
// Assigning to function/classes with an empty yields the desired formatting
a = function () {};
a = class {};

// Assigning to a function/class with a non-empty body groups the whole assignment expression.
a =
  function () {
    console.log(1);
  };

extraLongLeftHandSideLetsSeeHowPrettierBreaksThatAcrossMultipleLinesStillNotLongEnoughMustBeLongerAndLora =
  function () {
    console.log(1);
  };

a =
  function extraLongLeftHandSideLetsSeeHowPrettierBreaksThatAcrossMultipleLinesStillNotLongEnoughMustBeLongerAndara() {
    console.log(1);
  };

a =
  class {
    prop;
  };

// Function with non-empty body is printed on the next line
A.prototype.f =
  function () {
    console.log(1);
  };
```

## Expected

Rome formats `class` and `function` expressions on a single line if there name fits on a single line.

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

Fixes #2417

This actually fixes a loads of stuff, although it would be to know from @leops if my change actually makes sense.

I know that not having a soft line break has a small repercussion on how comments are printed.

## Test Plan

Updated snapshots and checked if they make sense

### CODE REVIEW

File: crates/rome_js_formatter/tests/specs/prettier/js/assignment/binaryish.js.snap [hunk 1]
```diff
@@ -38,8 +37,7 @@ const computedDescriptionLines2 =
   (focused && !loading && descriptionLinesFocused) || // comment
   descriptionLines; // comment
 
-computedDescriptionLines =
-  (focused && !loading && descriptionLinesFocused) || descriptionLines;
+computedDescriptionLines = (focused && !loading && descriptionLinesFocused) || descriptionLines;
```
Comment (MichaReiser) [inline crates/rome_js_formatter/tests/specs/prettier/js/assignment/binaryish.js.snap:40]: This doesn't look right

**Review by MichaReiser — COMMENTED** [review]
Would you mind adding a few tests for the cases mentioned in the issues and also testing other edge cases (e.g. all other expressions should begin on a new line)

File: crates/rome_js_formatter/tests/specs/prettier/js/assignment/issue-1966.js.snap [hunk 2]
```diff
@@ -18,8 +17,7 @@ this.isaverylongmethodexpression.withmultiplelevels = this.isanotherverylongexpr
 ```js
 const aVeryLongNameThatGoesOnAndOn = this.someOtherObject.someOtherNestedObject.someLongFunctionName();
 
-this.someObject.someOtherNestedObject =
-  this.someOtherObject.whyNotNestAnotherOne.someLongFunctionName();
+this.someObject.someOtherNestedObject = this.someOtherObject.whyNotNestAnotherOne.someLongFunctionName();
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/rome_js_formatter/tests/specs/prettier/js/assignment/issue-1966.js.snap:20]: The issue here is that member expressions don't implement grouping.

File: crates/rome_js_formatter/src/js/expressions/assignment_expression.rs [hunk 3]
```diff
@@ -16,11 +14,19 @@ impl ToFormatElement for JsAssignmentExpression {
...
         } = self.as_fields();
 
+        let right = right?;
+
+        let right = if matches!(right, JsAnyExpression::JsAssignmentExpression(_)) {
+            soft_line_indent_or_space(right.format(formatter)?)
+        } else {
+            format_elements![space_token(), right.format(formatter)?]
+        };
+
```
Comment (MichaReiser) [inline crates/rome_js_formatter/src/js/expressions/assignment_expression.rs:24]: Nit: You can write this as 
```suggestion
        let right = right?;

        let right = if let JsAnyExpression::JsAnyAssignmentExpression(right)) = right?  {
            soft_line_indent_or_space(right.format(formatter)?)
        } else {
            format_elements![space_token(), right.format(formatter)?]
        };

```

**Review by MichaReiser — CHANGES_REQUESTED** [review]
While this seems to fix some of the issue, it also introduce many news. I suggest you to create the prettier report on main, then on your branch and then compare the reports to see the regressions. 

Some notable regressions:

* https://gist.github.com/MichaReiser/147a7c787f8cbaaa43c92dd8ceb3060f#jsassignment-commentsnumberjs
* https://gist.github.com/MichaReiser/147a7c787f8cbaaa43c92dd8ceb3060f#jsassignment-commentsstringjs
* https://gist.github.com/MichaReiser/147a7c787f8cbaaa43c92dd8ceb3060f#jsassignmentbinaryishjs
* https://gist.github.com/MichaReiser/147a7c787f8cbaaa43c92dd8ceb3060f#jsassignmentissue-2482-1js
* https://gist.github.com/MichaReiser/147a7c787f8cbaaa43c92dd8ceb3060f#jsassignmentissue-3819js

I believe the main issue is that your condition, testing for `AssignmentExpression` only is too open. You probably must check if the right hand side is a `FunctionExpression`, `ClassExpression`, `ArrayLIteral` or `ObjectLiteral` instead. 

Overall, please add new tests covering these cases. Also tests that verify the logic if either the left hand or right hand sides are very long, and potentially nested.

File: crates/rome_js_formatter/src/js/expressions/assignment_expression.rs (hunk 3 above)
Comment (ematipico) [reply crates/rome_js_formatter/src/js/expressions/assignment_expression.rs:24]: No I can't because `right` gets moved and can't be inside the `else` branch

File: crates/rome_js_formatter/tests/specs/prettier/js/assignment/binaryish.js.snap (hunk 1 above)
Comment (ematipico) [reply crates/rome_js_formatter/tests/specs/prettier/js/assignment/binaryish.js.snap:40]: What should we expect instead?

File: crates/rome_js_formatter/tests/specs/prettier/js/assignment/issue-1966.js.snap (hunk 2 above)
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_js_formatter/tests/specs/prettier/js/assignment/issue-1966.js.snap:20]: I think this should be fixed in a separate PR because it's not related to assignment expression

**Comment by ematipico** [comment]
> and potentially nested.

I will try but I would prefer doing it in a different PR because it involves a different logic. It's also tracked in a different PR which make things better.

**Comment by MichaReiser** [comment]
> > and potentially nested.
> 
> I will try but I would prefer doing it in a different PR because it involves a different logic. It's also tracked in a different PR which make things better.

I think we should at least have tests to see how your PR changes the nesting. Ideally, create the tests first, then make your changes.

File: crates/rome_js_formatter/tests/specs/prettier/js/assignment/binaryish.js.snap (hunk 1 above)
Comment (ematipico) [reply crates/rome_js_formatter/tests/specs/prettier/js/assignment/binaryish.js.snap:40]: I just double checked. This is correct for the right-hand part of the `JsAssignmentExpression`. The parenthesized expression should de responsible of doing the breaking, which is not covered in this PR

**Comment by ematipico** [comment]
Closing. I am hitting a limitation of the printer, which requires a different fix.
